src/plugins/autofill/autofill.js

```diff
@@ -231,47 +234,77 @@ export class Autofill extends BasePlugin {
 
     if (startOfDragCoords && startOfDragCoords.row > -1 && startOfDragCoords.col > -1) {
       const selectionData = this.getSelectionData();
-      const beforeAutofillHook = this.hot.runHooks('beforeAutofill', startOfDragCoords, endOfDragCoords, selectionData);
 
-      if (beforeAutofillHook === false) {
+      const sourceRange = {
+        from: {
+          row: Math.min(selectionRangeLast.from.row, selectionRangeLast.to.row),
+          col: Math.min(selectionRangeLast.from.col, selectionRangeLast.to.col)
+        },
+        to: {
+          row: Math.max(selectionRangeLast.from.row, selectionRangeLast.to.row),
+          col: Math.max(selectionRangeLast.from.col, selectionRangeLast.to.col)
+        }
+      };
+
+      const targetRange = {
+        from: {
+          row: Math.min(startOfDragCoords.row, endOfDragCoords.row),
+          col: Math.min(startOfDragCoords.col, endOfDragCoords.col)
+        },
+        to: {
+          row: Math.max(startOfDragCoords.row, endOfDragCoords.row),
+          col: Math.max(startOfDragCoords.col, endOfDragCoords.col)
+        }
+      };
+
+      const beforeAutofillHookResult = this.hot.runHooks(
+        'beforeAutofill',
+        selectionData,
+        sourceRange,
+        targetRange,
+        directionOfDrag
+      );
+
+      if (beforeAutofillHookResult === false) {
         this.hot.selection.highlight.getFill().clear();
         this.hot.render();
 
         return false;
       }
 
       const deltas = getDeltas(startOfDragCoords, endOfDragCoords, selectionData, directionOfDrag);
-      let fillData = selectionData;
 
-      if (['up', 'left'].indexOf(directionOfDrag) > -1) {
-        fillData = [];
+      let fillData = beforeAutofillHookResult;
+      const res = beforeAutofillHookResult;
 
-        let dragLength = null;
-        let fillOffset = null;
+      if (['up', 'left'].indexOf(directionOfDrag) > -1 && res.length > 0 && res[0].length > 0) {
```

Comment:
I have doubts about `res[0].length > 0`. What if the selection was:
```
[
  [],
  ['a', 'b'],
  ['c', 'd'],
]
```
autofill should work correctly, right?

Reply:
@wojciechczerniak If empty cells always contain `null` (?) that's an edge case and a half, but good catch, the condition should be inverted anyway. https://github.com/handsontable/handsontable/pull/7987/commits/335899fd7b7091706ffc7a531c77ad367e12848b

Reply:
You're right, it's always `null`. We're safe here